trace/tracestate/tracestate.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracestate
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+)
+
+const (
+	keyMaxSize       = 256
+	valueMaxSize     = 256
+	maxKeyValuePairs = 32
+)
+
+const (
+	keyWithoutVendorFormat = `[a-z][_0-9a-z\-\*\/]{0,255}`
+	keyWithVendorFormat    = `[a-z][_0-9a-z\-\*\/]{0,240}@[a-z][_0-9a-z\-\*\/]{0,13}`
+	keyFormat              = `(` + keyWithoutVendorFormat + `)|(` + keyWithVendorFormat + `)`
+	valueFormat            = `[\x20-\x2b\x2d-\x3c\x3e-\x7e]{0,255}[\x21-\x2b\x2d-\x3c\x3e-\x7e]`
+)
+
+var keyValidationRegExp = regexp.MustCompile(`^(` + keyFormat + `)$`)
+var valueValidationRegExp = regexp.MustCompile(`^(` + valueFormat + `)$`)
+
+// Tracestate represent tracing-system specific context in a list of key-value pairs. Tracestate allows different
+// vendors propagate additional information and inter-operate with their legacy Id formats.
+type Tracestate struct {
+	tracestateEntries []*TracestateEntry
+}
+
+// TracestateEntry represent one key-value pair in a list of key-value pair of Tracestate.
+// Key is an opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+// and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+// forward slashes /.
+//
+// Value is an opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+// range 0x20 to 0x7E) except comma , and =.
+type TracestateEntry struct {
```

Comment:
Do we need this type? We can represent everything as <string, string>.

In Go, introduction of new types are not encouraged if builtin types are sufficient. You can write the NewFromTracestateEntries as:

```
NewFromKeyVals(map[string]string) (*Tracestate, error)
```

Reply:
with map[string]string order is not maintained.

Reply:
It was useful to know that order matters, thanks.

In Go, we don't stutter. Let's rename TracestateEntry to just Entry. Then it will look like tracestate.Entry at the call site.